src/browser/BrowserAction.cpp

```diff
@@ -0,0 +1,534 @@
+    message["entries"] = arr;
+    message["nonce"] = nonce;
+
+    QJsonObject response;
+    response["action"] = action;
+    response["message"] = encryptMessage(message, nonce);
+    response["nonce"] = nonce;
+
+    return response;
+}
+
+const QJsonObject BrowserAction::handleSetLogin(const QJsonObject& json, const QString& action)
+{
+    const QString hash = getDataBaseHash();
+    const QString nonce = json.value("nonce").toString();
+    const QString encrypted = json.value("message").toString();
+
+    QMutexLocker locker(&m_mutex);
+    if (!m_associated) {
+        return getErrorReply(action, ERROR_KEEPASS_ASSOCIATION_FAILED);
+    }
+
+    const QJsonObject decrypted = decryptMessage(encrypted, nonce, action);
+    if (!decrypted.isEmpty()) {
+        const QString url = decrypted.value("url").toString();
+        if (url.isEmpty()) {
+            return getErrorReply(action, ERROR_KEEPASS_NO_URL_PROVIDED);
+        } else {
+            const QString id = decrypted.value("id").toString();
+            const QString login = decrypted.value("login").toString();
+            const QString password = decrypted.value("password").toString();
+            const QString submitUrl = decrypted.value("submitUrl").toString();
+            const QString uuid = decrypted.value("uuid").toString();
+            const QString realm = ""; // ?
+
+            if (uuid.isEmpty()) {
+                m_browserService.addEntry(id, login, password, url, submitUrl, realm);
+            } else {
+                m_browserService.updateEntry(id, uuid, login, password, url);
+            }
+
+            QJsonObject message;
+            message["count"] = QJsonValue::Null;
+            message["entries"] = QJsonValue::Null;
+            message["error"] = "";
+            message["hash"] = hash;
+            message["version"] = KEEPASSX_VERSION;
+            message["success"] = "true";
+            message["nonce"] = nonce;
+
+            QJsonObject response;
+            response["action"] = action;
+            response["message"] = encryptMessage(message, nonce);
+            response["nonce"] = nonce;
+
+            return response;
+        }
+    }
+    return getErrorReply(action, ERROR_KEEPASS_CANNOT_DECRYPT_MESSAGE);
+}
+
+const QJsonObject BrowserAction::handleLockDatabase(const QJsonObject& json, const QString& action)
+{
+    const QString hash = getDataBaseHash();
+    const QString nonce = json.value("nonce").toString();
+    const QString encrypted = json.value("message").toString();
+    const QJsonObject decrypted = decryptMessage(encrypted, nonce, action);
+
+    if (!decrypted.isEmpty()) {
+        QString command = decrypted.value("action").toString();
+        if (!hash.isEmpty() && command.compare("lock-database", Qt::CaseSensitive) == 0) {
+            QMutexLocker locker(&m_mutex);
+            m_browserService.lockDatabase();
+            return getErrorReply(action, ERROR_KEEPASS_DATABASE_NOT_OPENED);
```

Comment:
This is only because the browser extension locks the database when that error is seen. I'll change the behaviour to a more logical one instead of this dirty hack.

Reply:
I know. But since we develop a custom browser extension anyway, we don't need to carry over dirty hacks from KeePassHTTP.